Lib/asyncio/queues.py

```diff
@@ -57,8 +64,9 @@ def _put(self, item):
 
     def _wakeup_next(self, waiters):
         # Wake up the next waiter (if any) that isn't cancelled.
-        while waiters:
-            waiter = waiters.popleft()
+        # from gh-83055: don't remove future from waiters,
+        # do it in get() or put() when `waiter` wake up.
+        for waiter in waiters:
```

Comment:
A problem here is that if there are many waiters, this changes from O(1) to O(N), as each waiter has to loop over all the already-done waiters to find the next one that's not yet done.

Reply:
Sorry for ignoring this complexity issue.
There's a way to avoid modifying the `_wakeup_next function and preserve the O(1). We need to add a new private attribute to count the number of awake tasks in 'transit' between the waiters and the internal queue. If this counter is not 0, a new element must wait for its time.